addons/stock_picking_batch/models/stock_picking_batch.py

```diff
@@ -17,48 +17,121 @@ class StockPickingBatch(models.Model):
+        domain="[('id', 'in', allowed_picking_ids)]", check_company=True,
+        states={'draft': [('readonly', False)], 'in_progress': [('readonly', False)]},
         help='List of transfers associated to this batch')
+    show_check_availability = fields.Boolean(
+        compute='_compute_move_ids',
+        help='Technical field used to compute whether the check availability button should be shown.')
+    allowed_picking_ids = fields.One2many('stock.picking', compute='_compute_allowed_picking_ids')
+    move_ids = fields.One2many(
+        'stock.move', string="Stock moves", compute='_compute_move_ids')
+    move_line_ids = fields.One2many(
+        'stock.move.line', 'batch_id', string='Stock move lines')
     state = fields.Selection([
         ('draft', 'Draft'),
         ('in_progress', 'In progress'),
         ('done', 'Done'),
         ('cancel', 'Cancelled')], default='draft',
         copy=False, tracking=True, required=True, readonly=True)
+    picking_type_id = fields.Many2one(
+        'stock.picking.type', 'Operation Type', check_company=True, copy=False,
+        readonly=True, states={'draft': [('readonly', False)]})
 
+    @api.depends('company_id', 'picking_type_id')
+    def _compute_allowed_picking_ids(self):
+        allowed_picking_states = ['waiting', 'confirmed', 'assigned']
+        for batch in self:
+            domain_states = list(allowed_picking_states)
+            # Allows to add draft pickings only if batch is in draft as well.
+            if batch.state == 'draft':
+                domain_states.append('draft')
+            domain = [
+                ('company_id', '=', batch.company_id.id),
+                ('state', 'in', domain_states),
+                '|',
+                ('batch_id', '=', False),
+                ('batch_id', '=', batch.id),
+            ]
+            if batch.picking_type_id:
+                domain += [('picking_type_id', '=', batch.picking_type_id.id)]
+            allowed_pickings = self.env['stock.picking'].search_read(domain, ['id'])
+            allowed_picking_ids = [picking['id'] for picking in allowed_pickings]
+            batch.allowed_picking_ids = [(6, False, allowed_picking_ids)]
+
+    @api.depends('picking_ids', 'picking_ids.move_lines')
+    def _compute_move_ids(self):
+        for batch in self:
+            batch.move_ids = batch.picking_ids.move_lines
+            batch.show_check_availability = any(m.state != 'assigned' for m in batch.move_ids)
+
+
+    # -------------------------------------------------------------------------
+    # CRUD
+    # -------------------------------------------------------------------------
     @api.model
     def create(self, vals):
         if vals.get('name', '/') == '/':
             vals['name'] = self.env['ir.sequence'].next_by_code('picking.batch') or '/'
-        return super(StockPickingBatch, self).create(vals)
+        res = super().create(vals)
+        res._sanity_check()
```

Comment:
would rather do the check on picking. Since if you create a picking with a batch_id without going through the create on batch it could be allowed to add different picking_type.